domain/application/service/resource.go

```diff
@@ -36,34 +38,42 @@ type ResourceState interface {
 	// SetUnitResource sets the resource metadata for a specific unit.
 	SetUnitResource(ctx context.Context, config resource.SetUnitResourceArgs) (resource.SetUnitResourceResult, error)
 
-	// OpenResource returns the metadata for a resource and a reader for
-	// the resource.
-	OpenResource(ctx context.Context, resourceID resources.ID) (resource.Resource, io.ReadCloser, error)
-	// OpenApplicationResource?
+	// OpenApplicationResource returns the metadata for an application's resource.
+	OpenApplicationResource(ctx context.Context, resourceID resources.ID) (resource.Resource, error)
 
-	// OpenResourceForUniter returns the metadata for a resource and a reader
-	// for the resource. A unit resource is created to track the given unit and
-	// which resource its using.
-	OpenResourceForUniter(ctx context.Context, resourceID resources.ID, unitID coreunit.UUID) (resource.Resource, io.ReadCloser, error)
-	// OpenUnitResource?
+	// OpenUnitResource returns the metadata for a resource a. A unit resource is
+	// created to track the given unit and which resource its using.
+	OpenUnitResource(ctx context.Context, resourceID resources.ID, unitID coreunit.UUID) (resource.Resource, error)
 
 	// SetRepositoryResources sets the "polled" resources for the
 	// application to the provided values. The current data for this
 	// application/resource combination will be overwritten.
 	SetRepositoryResources(ctx context.Context, config resource.SetRepositoryResourcesArgs) error
 }
 
+// resourceStoreFn returns the appropriate object store for the
+// provided resource type.
+type resourceStoreFn func(context.Context, charmresource.Type) (objectstore.ObjectStore, error)
+
 // ResourceService provides the API for working with resources.
 type ResourceService struct {
-	st     ResourceState
-	logger logger.Logger
+	st               ResourceState
+	getResourceStore resourceStoreFn
+	logger           logger.Logger
 }
 
 // NewResourceService returns a new service reference wrapping the input state.
-func NewResourceService(st ResourceState, logger logger.Logger) *ResourceService {
+func NewResourceService(st ResourceState, objectStoreGetter objectstore.ModelObjectStoreGetter, logger logger.Logger) *ResourceService {
 	return &ResourceService{
-		st:     st,
-		logger: logger,
+		st: st,
+		getResourceStore: func(ctx context.Context, resourceType charmresource.Type) (objectstore.ObjectStore, error) {
```

Comment:
We should endow behaviour in types. 

The resource service doesn't need to know about the object store, it needs to know that given a path/config for a type (file or oci-image) and returns a io.Reader.

By passing the object store and instantiating the resource store here, we're dealing with a concrete type and we're coupling it. 

Even just moving this function out of the service would be a win.

Reply:
If we don't want to know about the object store in here then we will need a separate service for the oci-image store. The original plan was to just have a separate table for it, mirroring how we did it in mongo. If we want to inject this resource store as a dependency, we won't be able to access the application state from the OCI image store inside that dependency.

Because of the password issue in #18351 we may end up doing something else, but as far as I understand, if we're going to back the oci-image resource store with the database, we would need a separate service that would be hidden in this resource store injected as a dependency. 

Is my analysis here correct or is there a way round this?

Reply:
@SimonRichardson 

> We should endow behaviour in types.

That's what I'm attempting to start here for the service. The model object store is for File types, and a to be written store will wrap calls into the resource_oci_image_metadata_store table. I expect the interface returned by getResourceStore to change once that's done. This is scaffolding to get started.

As @Aflynn50 says, if we push this up a layer, we'll have to write a domain around interacting with resource_oci_image_metadata_store, which should essentially be pushing one service into another hidden in a wrapper.

> The resource service doesn't need to know about the object store, it needs to know that given a path/config for a type (file or oci-image) and returns a io.Reader.

Today SetResource in the old mongo state, also ["stores"](https://github.com/juju/juju/blob/ea9879763912db3de5f769c0532fc6ccbf615b23/state/resources.go#L971) the resource. The resource code is very convoluted. 

It'd take a bit to decouple what state does today, update resource documents and handles storage of the resource. For OpenResourceForUniter, there is a [special type which implements io.Read](https://github.com/juju/juju/blob/ea9879763912db3de5f769c0532fc6ccbf615b23/state/resources.go#L814) to allow for tracking download progress between the controller and the unit in the db. This progress can be seen by `juju resource <application|unit>` or anyone using the facade.

Reply:
I would delay this architectural decision and simply have a member for the the file object store getter.
When you have the dependency for the OCI storage, just add that as another getter.

It seems to me having a _method_ to choose between the implementations of object store getter rather than a _member func_ is more parsimonious.

Reply:
This is what I'm suggesting... https://github.com/hmlanigan/juju/pull/2

Reply:
He didn't want to make a big deal of it, and i wanted the function outside of the constructor (I wasn't going to block the pr though). What we've settled on is exactly what I was pushing for.